xarray/core/concat.py

```diff
@@ -568,38 +569,40 @@ def get_indexes(name):
                     yield PandasIndex(data, dim, coord_dtype=var.dtype)
 
     # create concatenation index, needed for later reindexing
-    concat_index = list(range(sum(concat_dim_lengths)))
+    concat_index = np.arange(sum(concat_dim_lengths))
+    concat_index_size = concat_index.size
+    variable_index_mask = np.ones(concat_index_size, dtype=bool)
 
     # stack up each variable and/or index to fill-out the dataset (in order)
     # n.b. this loop preserves variable order, needed for groupby.
+    ndatasets = len(datasets)
     for name in vars_order:
         if name in concat_over and name not in result_indexes:
             variables = []
-            variable_index = []
+            mask = variable_index_mask.copy()
             var_concat_dim_length = []
             for i, ds in enumerate(datasets):
                 if name in ds.variables:
                     variables.append(ds[name].variable)
-                    # add to variable index, needed for reindexing
-                    var_idx = [
-                        sum(concat_dim_lengths[:i]) + k
-                        for k in range(concat_dim_lengths[i])
-                    ]
-                    variable_index.extend(var_idx)
-                    var_concat_dim_length.append(len(var_idx))
```

Comment:
This was the major bottleneck.

Reply:
Great performance improvement, @Illviljan! It's really helpful to see how the number of datasets/variables effect the performance and how to reduce that drop. I hope I'll keep attention for these issues in future contributions. We've already talked about benchmarks for large number of datasets/variables back in #7400 (https://github.com/pydata/xarray/pull/7400#issuecomment-1378321206), great to have it now.

But seeing all this, I'm wondering if it would be nice to make this missing variable reindexing behaviour available as opt-in (and not by default). So people could gain even more performance if they know their datasets are well formed without missing variables (especially for very large number of datasets).

Reply:
Can you add a comment describing the approach please? It still takes some thinking to interpret whats going on

Reply:
@dcherian Sure, I can add comments at the different locations.

Reply:
Sorry I meant @Illviljan

Reply:
I've added some comments and maybe some better variable names. Is it any better?